torch/_functorch/aot_autograd.py

```diff
@@ -1219,25 +1225,72 @@ def call_func_with_args(f, args, steal_args=False, disable_amp=False):
             del guard
     return out
 
-def aot_dispatch_base(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *, fw_metadata: ViewAndMutationMeta):
+@dataclasses.dataclass
+class SubclassCreationMeta:
+    """
+    Used for AOTDispatch: The graph that we'd like to trace out contains flat tensor inputs,
+    And this dataclass gives us the information we need to reconstruct a tensor subclass
+    from our flat inputs.
+    """
+
+    creation_lambda: Callable[..., torch.Tensor]
+    flat_tensor_start_idx: int
+    arg_count: int
+
+@dataclasses.dataclass
+class TensorToSubclassInfo:
+    """
+    Used for AOTDispatch: Tells us globally how to convert a sequence
+    of dense tensor arguments into subclasses
+    """
+
+    wrap_info: Sequence[Union[int, SubclassCreationMeta]]
+    # Needed by the partitioner to know how many flat
+    # outputs there are in the fw graph
+    meta: ViewAndMutationMeta
+    # Needed by the partitioner to pass in flat joint inputs
+    maybe_joint_inputs: Sequence[Any]
+    # Needed by the runtime wrapper for the fw,
+    # so it can properly wrap dense tensor outputs back into subclasses
+    fw_output_infos: List[Union[int, SubclassCreationMeta]]
+    # Needed by the runtime wrapper for the bw,
+    # so it can properly wrap dense tensor outputs back into subclasses
+    bw_output_infos: List[Union[int, SubclassCreationMeta]]
+    # Needed during the runtime wrapper for the bw,
+    # so we can error out if we find that the grad_outputs
+    # have different subclass types than what we traced with
+    fw_output_types: List[type]
+
+
+def aot_dispatch_base(
+    flat_fn,
+    flat_args: List[Tensor],
+    aot_config: AOTConfig,
+    *,
+    fw_metadata: ViewAndMutationMeta,
+    # inner_fn takes in the joint function that we want to trace, and produces a flat FX graph on dense tensors
+    inner_fn: Callable[[Callable, Sequence[Any], AOTConfig, ViewAndMutationMeta, bool], torch.fx.GraphModule],
+    # Produes a runtime wrapper that wraps the compiled function
+    # It provides the glue from (subclass_inp -> dense_inp -> dense_out -> subclass_out)
+    inner_runtime_wrapper: Callable[[Callable, List[Union[int, SubclassCreationMeta]]], Callable],
```

Comment:
Callable in Callable in Python is too much. Separate this out into a protocol with __call__. This will also let you specify kwargs if that helps